### ISSUE TITLE
Private declaration around F# record or DU contents acts as internal declaration.

### ISSUE THREAD

**Issue by bryanedds** [Microsoft/visualfsharp#984]
It took me a while to notice this, but apparently declaring a record like so -
```
type Record =
    private
        { Field : int }
```
Does not make the field private. Instead, it makes the field internal to the module. I don't think this used to be the case, but maybe it has always been. It's especially troubling because my general approach to programming in F# using abstract data types relies on this higher-degree of encapsulation. In fact, I did an entire talk on the subject not too long ago here - https://vimeo.com/128464151

The same seems to be the case with discriminated unions -
```
type Union =
    private
        | Case
```
Privatizing union contents is also important since it's a common way to implement mocking systems like so -
```
type System =
    private
        | Mock
        | Implementation of string * int * bool
```
The only places that should have access to these privatized fields / sub-types is within the scope of the type's definition and from a module with the same name in the same namespace using the ```[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]``` attribute.

Let's please come up with a plan to fix this bug so that F# can continue to give optimal support for abstract data types, a fundamental and arguably optimal implementation and design technique in F# as used here - https://github.com/bryanedds/Prime - and in programming generally as here - https://github.com/bryanedds/ax

**Comment by jacqueline-homan** [comment]
I remember that talk, Bryan. As an aside, your description of OOP as "the
land of nouns" is how I now explain the difference between OOP and FP. Tip
of the hat to you for that one :) That was a very good talk, BTW.

On Thu, Feb 25, 2016 at 3:54 PM, Bryan Edds <notifications@github.com>
wrote:

> It took me a while to notice this, but apparently declaring a record like
> so -
>
> type Record =
> private
> { Field : int }
>
> Does not make the field private. Instead, it makes the field internal to
> the module. I don't think this used to be the case, but maybe it has always
> been. It's especially troubling because my general approach to programming
> in F# using abstract data types relies on this higher-degree of
> encapsulation. In fact, I did an entire talk on the subject not too long
> ago here - https://vimeo.com/128464151
>
> The same seems to be the case with discriminated unions -
>
> type Union =
> private
> | Case
>
> Privatizing union contents is also important since it's a common way to
> implement mocking systems like so -
>
> type System =
> private
> | Mock
> | Implementation of string * int * bool
>
> The only places that should have access to these privatized fields /
> sub-types is within the scope of the type's definition and from a module
> with the same name in the same namespace using the [] attribute.
>
> Let's please come up with a plan to fix this bug so that F# can continue
> to give optimal support for abstract data types, a fundamental and arguably
> optimal implementation and design technique in F# and in programming
> generally.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Microsoft/visualfsharp/issues/984>.
>



-- 



[image: --]
Jacqueline Homan
[image: http://]about.me/jacqueline_homan
<http://about.me/jacqueline_homan>

**Comment by FunctionalFirst** [comment]
@bryanedds `private` in this context has always meant "private to the module." See [this stackoverflow answer](http://stackoverflow.com/a/14462045/162396).

**Comment by bryanedds** [comment]
So then we need a new modifier, ```really_private```?

**Comment by dsyme** [comment]
Yes, this is according to spec.  Please add any suggestions for additions to fslang.uservoice.com

**Comment by bryanedds** [comment]
Actually, in that respect the standard doesn't apply because I don't define the type within a module. It's defined in a namespace.

**Comment by bryanedds** [comment]
Fine, I'll just open it in another issue.

Don't close my issues so quickly without the minimum of discussion please.

**Comment by dsyme** [comment]
@bryanedds Thanks. The spec can be clarified to indicate that it applies to namespaces too.

**Comment by bryanedds** [comment]
Okay, that's reasonable resolution. I'll close the related issue.

**Comment by dsyme** [comment]
@bryanedds Ah I see, or at least think I do, I suspect you're hitting the case where two separate files contribute to the same namespace.  There is indeed a change we want to make here, see https://fslang.uservoice.com/forums/245727-f-language/suggestions/6987069-restrict-private-for-items-in-namespaces-to-mean  which has been approved-in-principle - we would like to emit a warning in this case.

If it helps, the behavior for private type representations in modules is inherited from OCaml (though not namespaces). In F# 1.0 compat with the core of OCaml was more of an issue than it is now, OCaml has no members on types (and I know you don't like/use these in your approach to ADTs).

BTW I believe you can get the behavior you want with an AutoOpen module to scope the availability of the representation though I can see that may be too much ceremony to be satisfactory for you.

```
    [<AutoOpen>]
    module M = 
      type Record =
        private
            { Field : int }
```

**Comment by bryanedds** [comment]
Thx for following up!

I think that both fixes part of the issue, but then created a new one.

For example, I'll often define the types in one file and the functions in another like so -

https://github.com/bryanedds/Prime/blob/master/Prime/Prime/Prime/EventTypes.fs#L76-L80

https://github.com/bryanedds/Prime/blob/master/Prime/Prime/Prime/EventModules.fs#L16-L52

This is idiomatic for abstract data types in F# so far. However, this will be broken unless we add a special exception to the new proposed rule - that of - "... except where functions are defined in a module with the same name as the type and in the same namespace". Which is fine, because its nearly what I proposed above.

So the full proposal would be - 

Restrict "private" for items in namespaces to mean "private to the namespace declaration group" except where functions are defined in a module with the same name as the type and in the same namespace.

Should I add a comment to the feature request here - https://fslang.uservoice.com/forums/245727-f-language/suggestions/6987069-restrict-private-for-items-in-namespaces-to-mean

Or should I added a new one with it referenced?
